From 48659c3c56d357cb58c2c0483b1f815ad046d211 Mon Sep 17 00:00:00 2001
From: Todd Lipcon <todd@apache.org>
Date: Mon, 14 Jan 2013 21:34:22 +0000
Subject: [PATCH 0854/1357] HDFS-4400. DFSInputStream#getBlockReader: last retries should ignore the cache. Contributed by Colin Patrick McCabe.

git-svn-id: https://svn.apache.org/repos/asf/hadoop/common/branches/HDFS-347@1433144 13f79535-47bb-0310-9956-ffa450edef68
(cherry picked from commit 681737e78ba0ce574b92ff0ef3bd1794492af27e)
---
 .../hadoop-hdfs/CHANGES.HDFS-347.txt               |    2 ++
 .../org/apache/hadoop/hdfs/DFSInputStream.java     |    2 +-
 2 files changed, 3 insertions(+), 1 deletions(-)

diff --git a/hadoop-hdfs-project/hadoop-hdfs/CHANGES.HDFS-347.txt b/hadoop-hdfs-project/hadoop-hdfs/CHANGES.HDFS-347.txt
index ad9e4ab..1289d9e 100644
--- a/hadoop-hdfs-project/hadoop-hdfs/CHANGES.HDFS-347.txt
+++ b/hadoop-hdfs-project/hadoop-hdfs/CHANGES.HDFS-347.txt
@@ -17,3 +17,5 @@ HDFS-4388. DomainSocket should throw AsynchronousCloseException when appropriate
 HDFS-4390. Bypass UNIX domain socket unit tests when they cannot be run.
 (Colin Patrick McCabe via todd)
 
+HDFS-4400. DFSInputStream#getBlockReader: last retries should ignore the cache
+(Colin Patrick McCabe via todd)
diff --git a/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/DFSInputStream.java b/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/DFSInputStream.java
index c3db205..773195f 100644
--- a/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/DFSInputStream.java
+++ b/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/DFSInputStream.java
@@ -941,7 +941,7 @@ public class DFSInputStream extends FSInputStream implements ByteBufferReadable
     // equivalent to declaring the DataNode bad.
     boolean triedNonDomainSocketReader = false;
     for (int retries = 0;
-          retries < nCachedConnRetry && (!triedNonDomainSocketReader);
+          retries < nCachedConnRetry || (!triedNonDomainSocketReader);
           ++retries) {
       Peer peer = null;
       if (retries < nCachedConnRetry) {
-- 
1.7.0.4

